[docs] guide for sending push notifications directly through APNS and FCM (#9935)
expo/expo: docs/pages/push-notifications/sending-notifications-custom.md

---

> title: Sending Notifications with APNS & FCM

Reviewer: APNS -> APNs everywhere, that's just how Apple writes it

---

Communicating directly with APNS and FCM is much more complicated than sending notifications through [Expo's push notification service](../sending-notifications/), so you should only use this feature if you're prepared to undertake that complexity. Here are a few things you'll have to handle yourself if you choose to write your own server for FCM and APNS:

- Differentiating between native iOS & Android device tokens on your backend
- Twice the amount of backend code to write and maintain (code for communicating with FCM, and then code for communicating with APNS)
- Fetching responses from FCM and APNS to check if your notification went through, error handling, credentials management

> That being said, sometimes you need finer-grained control over your notifications, so communicating directly with APNS and FCM is necessary. Expo will never lock you in and force you to use our services, so the `expo-notifications` API is push service agnostic (you can use it with any push notification service).

Reviewer: ```suggestion
That being said, sometimes you need finer-grained control over your notifications, so communicating directly with APNS and FCM is necessary. The Expo platform does not lock you into using Expo's application services, and the `expo-notifications` API is push-service agnostic (you can use it with any push notification service).
```

I think it comes across better to explain that the platform is open.

## How do I write my own APNS & FCM servers?

> Before we begin communicating directly with APNS & FCM, there is one client-side change you'll need to make in your app. When using Expo's notification service, you collect the `ExponentPushToken` with [`getExpoPushTokenAsync`](../../versions/latest/sdk/notifications/#getexpopushtokenasyncoptions-expotokenoptions-expopushtoken). Now that you're bypassing Expo, you'll need to collect the native device token instead with [`getDevicePushTokenAsync`](../../versions/latest/sdk/notifications/#getdevicepushtokenasync-devicepushtoken).

Reviewer: ```suggestion
Before we begin communicating directly with APNS & FCM, there is one client-side change you'll need to make in your app. When using Expo's notification service, you collect the `ExponentPushToken` with [`getExpoPushTokenAsync`](../../versions/latest/sdk/notifications/#getexpopushtokenasyncoptions-expotokenoptions-expopushtoken). Now that you're not using Expo's notification service, you'll need to collect the native device token instead with [`getDevicePushTokenAsync`](../../versions/latest/sdk/notifications/#getdevicepushtokenasync-devicepushtoken).
```

```diff
import * as Notifications from 'expo-notifications';
...
- const token = (await Notifications.getExpoPushTokenAsync()).data;
+ const token = (await Notifications.getDevicePushTokenAsync()).data:
// send token off to your server
```

> Now that you have your native device token, we can start to implement our servers:

Reviewer: emphasize that these are for demonstration purposes, not production
```suggestion
Now that you have your native device token, we can start to implement our servers. Below are some very minimal examples of communicating with FCM and APNs:
```

## FCM Server

> This documentation is based off of [Google's documentation](https://firebase.google.com/docs/cloud-messaging/http-server-ref), and we're just going to cover the basics here to get you started.

Communicating with FCM is as simple as sending a POST request, but before sending or receiving any notifications, you'll need to follow the steps [in this documentation](../using-fcm/) to configure FCM (and get your `FCM-SERVER-KEY`).

```js
await fetch('https://fcm.googleapis.com/fcm/send', {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    Authorization: `key=<FCM-SERVER-KEY>`,
  },
  body: JSON.stringify({
    to: '<NATIVE-DEVICE-PUSH-TOKEN>',
    priority: 'normal',
    data: {
      experienceId: '@yourExpoUsername/yourProjectSlug',
      title: "\uD83D\uDCE7 You've got mail",
      message: 'Hello world! \uD83C\uDF10',
    },
  }),

> });

Reviewer: This is their legacy API, the v1 API looks different. I think they got rid of FCM API keys and instead do a credentials dance with a google application credentials JSON file.

We don't use the v1 API right now but are laying the groundwork if we ever decided we needed to and it made sense to spend time on it.

Author: Yeah, but think it's better in the case of this over-simplistic example to use the same credentials that a reader would already have from using the Expo notification service, rather than walk them through creating new credentials here, as well. We can add a note referencing the migration guide for the FCM v1 API though

Reviewer: Maybe just mention then that this is for FCM's legacy API.

Author: added - https://github.com/expo/expo/pull/9935/files#diff-987d102b1b6c85d2c4dd00cf253dd3f2R33

Communicating with APNS is a little more complicated than with FCM. There are some libraries that wrap all of this functionality into one or two function calls (like [`node-apn`](https://github.com/node-apn/node-apn)), but in this example we're going to use the minimum required libraries to give you a good understanding of what's happening.

### Authorization

The first thing you need before sending requests to APNS is permission to send notifications to your app, which is going to be granted via a JSON web token. This web token is generated using these iOS developer credentials:

- APN key (`.p8` file) associated with your app
- Key ID of the above `.p8` file
- Your Apple Team ID

```js
const jwt = require("jsonwebtoken");
const authorizationToken = jwt.sign(
  {
    iss: "YOUR-APPLE-TEAM-ID"
    iat: Math.round(new Date().getTime() / 1000),
  },
  fs.readFileSync("./path/to/appName_apns_key.p8", "utf8"),
  {
    header: {
      alg: "ES256",
      kid: "YOUR-P8-KEY-ID",
    },
  }
);
```

> ### HTTP2 Connection

Reviewer: ```suggestion
### HTTP/2 Connection
```

> Now that we have our `authorizationToken`, we can open up an HTTP2 connection to Apple's servers. In development, you'll want to send requests to `api.development.push.apple.com`, but in production requests should go to `api.push.apple.com`.

Reviewer: ```suggestion
Now that we have our `authorizationToken`, we can open up an HTTP/2 connection to Apple's servers. In development, you'll want to send requests to `api.development.push.apple.com`, but in production requests should go to `api.push.apple.com`.
```

Here's how to construct your request:

```js
const http2 = require('http2');

const client = http2.connect(
  IS_PRODUCTION ? 'https://api.push.apple.com' : 'https://api.sandbox.push.apple.com'
);

const request = client.request({
  ':method': 'POST',
  ':scheme': 'https',
  'apns-topic': 'YOUR-BUNDLE-IDENTIFIER',
  ':path': '/3/device/' + nativeDeviceToken, // This is the native device token you grabbed client-side
  authorization: `bearer ${authorizationToken}`, // This is the JSON web token we generated in the "Authorization" step above
});
request.setEncoding('utf8');

request.write(
  JSON.stringify({
    aps: {
      alert: {
        title: "\uD83D\uDCE7 You've got mail!",
        body: 'Hello world! \uD83C\uDF10',
      },
    },
    experienceId: '@yourExpoUsername/yourProjectSlug', // Required when testing in the Expo client app
  })
);
request.end();
```

> > This example is **very** minimal, and includes no error handling. For testing purposes, you should refer to [this example code](https://gist.github.com/cruzach/8535b87d9f6475a78df22b69cf86da03), instead.

Reviewer: ```suggestion
> This example is **very** minimal, and includes no error handling nor connection pooling. For testing purposes, you should refer to [this example code](https://gist.github.com/cruzach/8535b87d9f6475a78df22b69cf86da03), instead.
```

Also move the gist to an expo.fyi page so others can update it more easily.
